nvflare/private/fed/server/job_runner.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (c) 2021-2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os.path
+import threading
+import time
+
+from nvflare.apis.event_type import EventType
+from nvflare.apis.fl_component import FLComponent
+from nvflare.apis.fl_constant import RunProcessKey
+from nvflare.apis.fl_constant import SystemComponents, WorkspaceConstants
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.job_def import RunStatus, Job
+from nvflare.private.admin_defs import Message
+from nvflare.private.defs import RequestHeader, TrainingTopic
+from nvflare.private.fed.utils.fed_utils import deploy_app
+
+
+class JobRunner(FLComponent):
+
+    def __init__(self, workspace_root: str) -> None:
+        super().__init__()
+        self.workspace_root = workspace_root
+        self.ask_to_stop = False
+        self.scheduler = None
+        self.running_jobs = {}
+        self.lock = threading.Lock()
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.SYSTEM_START:
+            engine = fl_ctx.get_engine()
+            self.scheduler = engine.get_component(SystemComponents.JOB_SCHEDULER)
+
+    def _deploy_job(self, job: Job, sites: list, fl_ctx: FLContext) -> str:
+        """deploy the application to the list of participants
+
+        Args:
+            job: job to be deployed
+            sites: participating sites
+            fl_ctx: FLContext
+
+        Returns:
+
+        """
+        engine = fl_ctx.get_engine()
+        run_number = job.job_id
+        workspace = os.path.join(self.workspace_root, WorkspaceConstants.WORKSPACE_PREFIX + run_number)
+        count = 1
+        while os.path.exists(workspace):
+            run_number = run_number + "_" + str(count)
+            workspace = os.path.join(self.workspace_root, WorkspaceConstants.WORKSPACE_PREFIX + run_number)
+            count += 1
+
+        for app_name, participants in job.get_deployment().items():
+            app_data = job.get_application(app_name, fl_ctx)
```

Comment:
if here is the only usage of get_application, I would suggest we just use

`engine.get_component(SystemComponents.JOB_MANAGER).get_app`

directly.